Dimensions

shylux opened the pull request:
Original issue: https://github.com/4teamwork/pzbs.shop/issues/27

This PR adds a new option to specify the dimensions of a shop item, eg. the length of a piece of fabric.
There is a set of selectable dimensions as requested by our client, but it can be extended to any combination of units.
The dimensions are applied to the price of the shop item. For instance if you have fabric and the user selects 400mm the price will be multiplied by 400. So you have to specify the price by mm.

The dimensions have to be specified by the customer upon adding an item to the cart. They can be edited in the cart edit view. They are displayed in the order overview and specified in all emails. They also show up in the order manager and the csv export.

Also the "normal" view (contrary to the compact view) of a shop item has been removed. This has two reasons. First, the normal view was broken - you couldn't add an item to the cart. And second it would have been nigh impossible to add the dimensions without creating a ux atrocity. Keep in mind that the views are designed to work without javascript.
So the normal view has been replaced with the compact view.
![screen shot 2017-03-27 at 11 11 07](https://cloud.githubusercontent.com/assets/1375745/24393612/09cd85b8-1399-11e7-94ae-6115ff118bcb.png)
![screen shot 2017-03-27 at 11 12 06](https://cloud.githubusercontent.com/assets/1375745/24393615/0bd13f26-1399-11e7-8241-1630e1434bc5.png)
![screen shot 2017-03-27 at 11 13 09](https://cloud.githubusercontent.com/assets/1375745/24393619/0f240046-1399-11e7-88e6-2cbb39d26480.png)
![screen shot 2017-03-27 at 11 13 37](https://cloud.githubusercontent.com/assets/1375745/24393622/111bdbb2-1399-11e7-8067-075981201f98.png)

elioschmutz commented on ftw/shop/tests/test_shopitem_views.py:
```diff
@@ -48,15 +48,16 @@ def test_one_variation(self):
         pq_doc = pq(item_listing)
-        self.assertEquals(len(pq_doc("input[name=skuCode][value='11']")), 1)
+        self.assertEquals(len(pq_doc('select[name=var1choice] option')), 3)
+        self.assertEquals(len(pq_doc('select[name=var2choice] option')), 3)
```
Why do you change that?

elioschmutz commented on ftw/shop/tests/test_shopitem_views.py:
```diff
@@ -112,13 +113,9 @@ def test_variations_order(self):
-            and all_sku_codes.index(green_s) < all_sku_codes.index(blue_s))
+        self.assertEquals(
+            ['Red', 'Green', 'Blue'],
+            [op.text for op in pq_doc('select[name=var1choice] option')])
```
Why do you change that?

elioschmutz commented on ftw/shop/tests/test_shopitem_views.py:
```diff
@@ -129,14 +126,9 @@ def test_variations_order(self):
-
+        self.assertEquals(
+            ['Blue', 'Green', 'Red'],
+            [op.text for op in pq_doc('select[name=var1choice] option')])
```
Why do you change that?

elioschmutz commented on ftw/shop/browser/templates/order_view.pt:
```diff
@@ -99,6 +99,7 @@
         <thead>
           <tr>
             <th i18n:translate="cart_header_product">Product</th>
+            <th i18n:translate="cart_header_dimensions">Dimensions</th>
```
Do not include Dimensions if there are no shopitems with dimensions.

elioschmutz commented on ftw/shop/browser/templates/order_view.pt:
```diff
@@ -108,6 +109,7 @@
           <metal:block tal:repeat="item python:order.cartitems">
             <tr>
               <td tal:content="item/title"></td>
+              <td><span metal:use-macro="context/@@shopitem_macros/dimensions_display"></span></td>
```
Do not include Dimensions if there are no shopitems with dimensions.

elioschmutz commented on ftw/shop/tests/test_cart.py:
```diff
@@ -190,7 +191,7 @@ def test_checkout(self):
 
         order = omanager.getOrders()[0]
-        self.assertEquals(order.total, Decimal('7.15'))
+        self.assertEquals(order.total, Decimal('14.30'))
```
You just extend existing tests, but you added a new feature. The current testcases are very large, but perhaps you can add own testscases where you just test your feature here.
(this comment is for all the tests added in this commit)

shylux commented on ftw/shop/tests/test_category_view.py:
```diff
@@ -36,16 +36,17 @@ def test_one_variation(self):
-        self.assertEquals(len(pq_doc("input[name=skuCode][value=b11]")), 1)
-        self.assertEquals(len(pq_doc("input[name=skuCode][value=b22]")), 1)
+        self.assertEquals(pq_doc("#var-Hardcover td")[2].text, 'b11')
+        self.assertEquals(pq_doc("#var-Paperback td")[2].text, 'b22')
```
The default view changed from normal to compact so i had to update those tests.

shylux commented:
* Added description for dimensions field.
 * Only display the dimensions column if they are used. (Tests for checkout etc cover that)
 * Added explicit test for dimensions selection & edit.
 * **Move comment field to review step.** This is another requirement which wasn't originally included in this PR.

@elioschmutz

elioschmutz commented:
The following functions are still untestet:

- calc_item_total
- calc_vat
- has_specified_dimensions
- has_order_dimensions
- cart_has_dimensions
- validate_dimensions

You added some explicit dimension-tests implemented as a browser test. Thats good for an overall test. But to test the functions above, you should add functional-tests with all possible variations and all if/else cases. 

Some test examples:

```python
shopping_cart = IShoppingCart(obj, request)
self.assertEqual(15, shopping_cart.calc_item_total(15, 1, [1]), "Price should be respected in the calculation")
self.assertEqual(30, shopping_cart.calc_item_total(15, 2, [1]), "Quantity should be respected in the calculation")
self.assertEqual(30, shopping_cart.calc_item_total(15, 1, [2]), "Dimension should be respected in the calculation")
self.assertEqual(60, shopping_cart.calc_item_total(15, 1, [1, 2, 2]), "Multiple Dimensions should be respected in the calculation")
```

shylux commented:
@elioschmutz 
Changed description to clarify price calculation.
![screen shot 2017-04-04 at 10 48 34](https://cloud.githubusercontent.com/assets/1375745/24650635/0df3eff0-192b-11e7-932c-d71332abb200.png)

It should be clear enough on the item view:
![screen shot 2017-04-04 at 11 39 14](https://cloud.githubusercontent.com/assets/1375745/24650737/6fbdef6a-192b-11e7-9212-bff0db11e95b.png)

Added tests for dimension validation and new `calc_item_total` method. I didn't test the other methods because i think they are sufficiently covered by the integration tests.
